File: UleamBack/payment-service/src/payment/adapters/mock.adapter.ts
/**
 * Mock Adapter - Implementación simulada de proveedor de pago
 * 
 * Este adaptador es OBLIGATORIO para desarrollo y testing.
 * NO realiza cargos reales, solo simula transacciones exitosas.
 * 
 * Características:
 * - Genera transaction_id único con formato: mock_txn_{uuid sin guiones}
 * - Siempre retorna estado "completed"
 * - Retraso simulado de 500ms para emular latencia de red
 * - Útil para testing sin necesidad de credenciales reales
 */
import { Injectable } from '@nestjs/common';
import { v4 as uuidv4 } from 'uuid';
import {
  PaymentProvider,
  PaymentData,
  PaymentResult,
  PaymentStatus,
} from '../interfaces/payment-provider.interface';

@Injectable()
export class MockAdapter extends PaymentProvider {
  readonly name = 'mock';

  /**
   * Procesar un pago simulado
   * - Genera transaction_id único
   * - Siempre retorna "completed"
   * - Simula 500ms de latencia de red
   */
  async processPayment(data: PaymentData): Promise<PaymentResult> {
    // Simular latencia de red
    await this.delay(500);

    // Generar transaction ID único: mock_txn_{uuid sin guiones}
    const uuid = uuidv4().replace(/-/g, '');
    const transactionId = `mock_txn_${uuid}`;

    return {
      transactionId,
      status: PaymentStatus.COMPLETED,
      message: `Pago simulado exitoso por ${data.amount} ${data.currency}`,
      providerData: {
        mock: true,
        processedAt: new Date().toISOString(),
        reservaId: data.reservaId,
        usuarioId: data.usuarioId,
      },
    };
  }

  /**
   * Verificar estado de pago simulado
   * - Siempre retorna "completed" para cualquier transaction_id que comience con "mock_txn_"
   */
  async verifyPayment(transactionId: string): Promise<PaymentResult> {
    await this.delay(200);

    if (!transactionId.startsWith('mock_txn_')) {
      return {
        transactionId,
        status: PaymentStatus.FAILED,
        message: 'Transaction ID no pertenece a MockAdapter',
      };
    }

    return {
      transactionId,
      status: PaymentStatus.COMPLETED,
      message: 'Pago simulado verificado',
      providerData: {
        mock: true,
        verifiedAt: new Date().toISOString(),
      },
    };
  }

  /**
   * Reembolsar un pago simulado
   * - Siempre retorna "refunded" exitosamente
   */
  async refundPayment(transactionId: string): Promise<PaymentResult> {
    await this.delay(300);

    if (!transactionId.startsWith('mock_txn_')) {
      return {
        transactionId,
        status: PaymentStatus.FAILED,
        message: 'Transaction ID no pertenece a MockAdapter',
      };
    }

    return {
      transactionId,
      status: PaymentStatus.REFUNDED,
      message: 'Reembolso simulado exitoso',
      providerData: {
        mock: true,
        refundedAt: new Date().toISOString(),
      },
    };
  }

  /**
   * Simular latencia de red
   */
  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: UleamBack/payment-service/app/adapters/mock_adapter.py
"""
Mock Payment Adapter

Implementación de prueba del PaymentProvider que simula un proveedor de pago
sin hacer llamadas a APIs externas.

Ideal para:
- Desarrollo local sin credenciales reales
- Testing automatizado
- Demos y pruebas de concepto

Características:
- Auto-completa pagos instantáneamente
- Soporta todos los métodos del PaymentProvider
- No requiere configuración externa
"""

import uuid
import logging
from typing import Dict, Any, Optional
from datetime import datetime

from .base import PaymentProvider
from ..schemas.webhook import NormalizedWebhookEvent

logger = logging.getLogger(__name__)


class MockAdapter(PaymentProvider):
    """
    Adapter de prueba que simula un payment provider sin APIs externas.
    
    Comportamiento:
    - create_payment: Auto-completa el pago instantáneamente
    - get_payment_status: Retorna siempre "completed"
    - refund_payment: Simula reembolso exitoso
    - validate_webhook_signature: Siempre retorna True (sin validación)
    
    IMPORTANTE: Solo usar en development/testing, NUNCA en producción.
    """
    
    def __init__(self):
        """Inicializa el MockAdapter con almacenamiento en memoria"""
        self.payments: Dict[str, Dict[str, Any]] = {}
        self.refunds: Dict[str, Dict[str, Any]] = {}
        logger.info("MockAdapter initialized (in-memory storage)")
    
    def create_payment(
        self,
        amount: float,
        currency: str,
        metadata: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Simula la creación de un pago.
        
        El pago se completa instantáneamente para facilitar testing.
        En un provider real, el pago quedaría en estado "pending".
        
        Args:
            amount: Monto del pago
            currency: Moneda (USD, EUR, etc.)
            metadata: Datos adicionales
        
        Returns:
            Dict con los datos del pago simulado
        """
        # Generar ID único simulado
        payment_id = f"mock_{uuid.uuid4().hex[:12]}"
        
        # Simular datos del pago
        payment_data = {
            "payment_id": payment_id,
            "status": "completed",  # Mock auto-completa
            "amount": amount,
            "currency": currency.upper(),
            "metadata": metadata,
            "client_secret": f"mock_secret_{uuid.uuid4().hex[:8]}",
            "created_at": datetime.utcnow()
        }
        
        # Almacenar en memoria (simula BD del provider)
        self.payments[payment_id] = payment_data
        
        logger.info(
            f"Mock payment created: {payment_id} "
            f"(amount={amount} {currency}, auto-completed)"
        )
        
        return payment_data
    
    def get_payment_status(self, payment_id: str) -> str:
        """
        Consulta el estado de un pago simulado.
        
        Args:
            payment_id: ID del pago
        
        Returns:
            str: Estado del pago (siempre "completed" si existe)
        
        Raises:
            ValueError: Si el payment_id no existe
        """
        if payment_id not in self.payments:
            logger.warning(f"Payment not found in MockAdapter: {payment_id}")
            raise ValueError(f"Payment {payment_id} not found in mock storage")
        
        status = self.payments[payment_id]["status"]
        logger.debug(f"Mock payment status retrieved: {payment_id} -> {status}")
        
        return status
    
    def refund_payment(
        self,
        payment_id: str,
        amount: Optional[float] = None,
        reason: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Simula un reembolso de pago.
        
        Args:
            payment_id: ID del pago a reembolsar
            amount: Monto a reembolsar (None = reembolso total)
            reason: Motivo del reembolso
        
        Returns:
            Dict con los datos del reembolso simulado
        
        Raises:
            ValueError: Si el pago no existe o no puede ser reembolsado
        """
        if payment_id not in self.payments:
            raise ValueError(f"Payment {payment_id} not found")
        
        payment = self.payments[payment_id]
        
        if payment["status"] != "completed":
            raise ValueError(
                f"Payment {payment_id} cannot be refunded (status: {payment['status']})"
            )
        
        # Calcular monto del reembolso
        refund_amount = amount if amount is not None else payment["amount"]
        
        if refund_amount > payment["amount"]:
            raise ValueError(
                f"Refund amount ({refund_amount}) exceeds payment amount ({payment['amount']})"
            )
        
        # Generar ID del reembolso
        refund_id = f"mock_refund_{uuid.uuid4().hex[:12]}"
        
        # Datos del reembolso
        refund_data = {
            "refund_id": refund_id,
            "payment_id": payment_id,
            "amount": refund_amount,
            "reason": reason or "No reason provided",
            "status": "succeeded",
            "created_at": datetime.utcnow()
        }
        
        # Almacenar reembolso
        self.refunds[refund_id] = refund_data
        
        # Actualizar estado del pago
        payment["status"] = "refunded"
        payment["refunded_amount"] = refund_amount
        
        logger.info(
            f"Mock refund created: {refund_id} "
            f"(payment={payment_id}, amount={refund_amount})"
        )
        
        return refund_data
    
    def normalize_webhook(self, raw_data: Dict[str, Any]) -> NormalizedWebhookEvent:
        """
        Normaliza un webhook simulado.
        
        MockAdapter recibe webhooks ya en formato normalizado,
        por lo que solo necesita construir el objeto NormalizedWebhookEvent.
        
        Args:
            raw_data: Datos del webhook
        
        Returns:
            NormalizedWebhookEvent con estructura normalizada
        """
        # Mock adapter espera datos ya normalizados
        return NormalizedWebhookEvent(
            event_type=raw_data.get("event_type", "payment.success"),
            payment_id=raw_data["payment_id"],
            status=raw_data.get("status", "completed"),
            amount=raw_data["amount"],
            currency=raw_data["currency"],
            metadata=raw_data.get("metadata", {}),
            error_message=raw_data.get("error_message"),
            timestamp=raw_data.get("timestamp", datetime.utcnow())
        )
    
    def validate_webhook_signature(
        self,
        payload: bytes,
        signature: str,
        timestamp: Optional[str] = None
    ) -> bool:
        """
        Simula validación de firma de webhook.
        
        IMPORTANTE: MockAdapter acepta CUALQUIER firma para facilitar testing.
        En producción, NUNCA usar MockAdapter.
        
        Args:
            payload: Payload del webhook
            signature: Firma recibida
            timestamp: Timestamp del webhook
        
        Returns:
            bool: Siempre True (sin validación real)
        """
        logger.warning(
            "MockAdapter.validate_webhook_signature: "
            "No real validation (always returns True)"
        )
        return True
    
    def supports_refunds(self) -> bool:
        """MockAdapter soporta reembolsos simulados"""
        return True
    
    def get_webhook_events(self) -> list[str]:
        """Eventos de webhook soportados por MockAdapter"""
        return [
            "payment.success",
            "payment.failed",
            "payment.cancelled",
            "payment.refunded",
            "payment.pending"
        ]
    
    def get_payment_details(self, payment_id: str) -> Optional[Dict[str, Any]]:
        """
        Método auxiliar para obtener detalles completos del pago.
        
        Útil para debugging y testing.
        
        Args:
            payment_id: ID del pago
        
        Returns:
            Dict con todos los datos del pago o None si no existe
        """
        return self.payments.get(payment_id)
    
    def clear_storage(self) -> None:
        """
        Limpia el almacenamiento en memoria.
        
        Útil para testing cuando necesitas resetear el estado.
        """
        self.payments.clear()
        self.refunds.clear()
        logger.info("MockAdapter storage cleared")
    
    def get_storage_stats(self) -> Dict[str, int]:
        """
        Retorna estadísticas del almacenamiento.
        
        Returns:
            Dict con contadores de pagos y reembolsos
        """
        return {
            "total_payments": len(self.payments),
            "total_refunds": len(self.refunds),
            "completed_payments": sum(
                1 for p in self.payments.values() if p["status"] == "completed"
            ),
            "refunded_payments": sum(
                1 for p in self.payments.values() if p["status"] == "refunded"
            )
        }
